In [1]:
import os

import numpy as np
import pandas as pd

from IPython.display import display

DIR = r'c://downloads'

In [2]:
new_cancer_cases_by_state = pd.read_csv(os.path.join(DIR, 'new_cancer_cases_by_state.csv'), thousands = ',', index_col = 'State')
new_cancer_cases_by_state = new_cancer_cases_by_state.iloc[:-1, 1:]
display(new_cancer_cases_by_state)

Female Breast  Uterine Cervix  Colon & Rectum  \
State                                                              
Alabama                     3660           210.0            2350   
Alaska                       450             NaN             280   
Arizona                     4520           210.0            2560   
Arkansas                    2050           140.0            1500   
California                 26130          1550.0           13930   
Colorado                    3780           160.0            1720   
Connecticut                 3160           120.0            1650   
Delaware                     760             NaN             420   
Dist. of Columbia            430             NaN             250   
Florida                    15480           960.0           10230   
Georgia                     7010           420.0            3940   
Hawaii                      1090            60.0             700   
Idaho                       1100            50.0             610   
Illinois                    9230           470.0            5530   
Indiana                     4590           260.0            3020   
Iowa                        2320           100.0            1580   
Kansas                      2090           100.0            1120   
Kentucky                    3370           200.0            2170   
Louisiana                   3160           200.0            2270   
Maine                       1220            50.0             700   
Maryland                    4570           230.0            2500   
Massachusetts               5560           200.0            2800   
Michigan                    7660           340.0            4570   
Minnesota                   3820           130.0            2240   
Mississippi                 2130           140.0            1510   
Missouri                    4610           240.0            2970   
Montana                      860             NaN             500   
Nebraska                    1360            60.0             880   
Nevada                      1880           120.0            1290   
New Hampshire               1150             NaN             600   
New Jersey                  7290           380.0            4280   
New Mexico                  1450            80.0             830   
New York                   15230           850.0            8590   
North Carolina              7580           380.0            4230   
North Dakota                 510             NaN             350   
Ohio                        8710           400.0            5450   
Oklahoma                    2700           160.0            1760   
Oregon                      3320           130.0            1540   
Pennsylvania               10660           500.0            6790   
Rhode Island                 870             NaN             500   
South Carolina              3750           210.0            2200   
South Dakota                 600             NaN             410   
Tennessee                   4840           290.0            3030   
Texas                      16080          1140.0            9760   
Utah                        1440            60.0             650   
Vermont                      560             NaN             290   
Virginia                    6170           290.0            3280   
Washington                  5620           230.0            2670   
West Virginia               1350            90.0            1060   
Wisconsin                   4330           190.0            2520   
Wyoming                      420             NaN             240   

                   Uterine Corpus  Leukemia  Lung & Bronchus  \
State                                                          
Alabama                       650       690             4160   
Alaska                        100       100              430   
Arizona                       910       950             4280   
Arkansas                      400       480             2660   
California                   5

In [3]:
from scipy.stats import fisher_exact, chi2_contingency
from statsmodels.stats.multitest import multipletests

_, pval, _, expected_data = chi2_contingency(new_cancer_cases_by_state.dropna())
print('%.2f of the expected values are above 5, %.2f are above 1' % (np.mean(expected_data >= 5), np.mean(expected_data >= 1)))
print('Chi-squared p-value: %e' % pval)

1.00 of the expected values are above 5, 1.00 are above 1
Chi-squared p-value: 0.000000e+00


In [4]:
from ipywidgets import FloatProgress

test_cancer_types = []
test_states = []
test_RRs = []
test_pvals = []

progress_bar = FloatProgress(max = new_cancer_cases_by_state.size)
display(progress_bar)

for cancer_type in new_cancer_cases_by_state.columns:

    other_cancer_types = new_cancer_cases_by_state.loc[:, new_cancer_cases_by_state.columns != cancer_type].sum(axis = 1)
    other_cancer_types.name = 'Other'
    cancer_type_data = pd.concat([new_cancer_cases_by_state[cancer_type], other_cancer_types], axis = 1)

    for state in new_cancer_cases_by_state.index:
    
        progress_bar.value += 1
        
        other_states = cancer_type_data[other_cancer_types.index != state].sum()
        other_states.name = 'Other'
        data = pd.DataFrame([cancer_type_data.loc[state], other_states])
        
        if pd.isnull(data).any().any():
            continue
        
        state_risk = data.iloc[0, 0] / data.iloc[0, :].sum()
        other_states_risk = data.iloc[1, 0] / data.iloc[1, :].sum()
        RR = state_risk / other_states_risk
        _, pval = fisher_exact(data)
        
        test_cancer_types.append(cancer_type)
        test_states.append(state)
        test_RRs.append(RR)
        test_pvals.append(pval)
        
test_summary = pd.DataFrame({'cancer_type': test_cancer_types, 'state': test_states, 'RR': test_RRs, 'pval': test_pvals})

FloatProgress(value=0.0, max=510.0)

In [5]:
test_summary['significance'], test_summary['qval'], _, _ = multipletests(test_summary['pval'], method = 'bonferroni')

# Sorting by max{RR, 1 / RR}
test_summary = test_summary.assign(absolute_RR = test_summary['RR'].apply(lambda RR: max(RR, 1 / RR)))\
        .sort_values('absolute_RR', ascending = False).drop('absolute_RR', axis = 1)

display(test_summary[test_summary['significance']])

cancer_type              state        RR           pval  \
289       Lung & Bronchus               Utah  0.481162  2.097583e-117   
340  Melanoma of the Skin               Utah  1.691588   3.951260e-44   
297  Melanoma of the Skin             Alaska  0.606716   1.538582e-07   
304  Melanoma of the Skin  Dist. of Columbia  0.634412   7.395569e-06   
457       Urinary Bladder  Dist. of Columbia  0.646488   2.187035e-05   
339  Melanoma of the Skin              Texas  0.658463  8.247214e-152   
299  Melanoma of the Skin           Arkansas  0.664275   3.017647e-23   
71         Uterine Cervix          Minnesota  0.668900   1.202174e-06   
314  Melanoma of the Skin          Louisiana  0.683422   5.824079e-30   
86         Uterine Cervix              Texas  1.455509   8.820529e-31   
332  Melanoma of the Skin           Oklahoma  0.707005   8.660023e-22   
69         Uterine Cervix      Massachusetts  0.709178   3.739552e-07   
161        Uterine Corpus          Louisiana  0.711514   2.475344e-17   
343  Melanoma of the Skin         Washington  1.390024   1.219884e-55   
333  Melanoma of the Skin             Oregon  1.387356   5.588858e-34   
167        Uterine Corpus        Mississippi  0.733301   3.373932e-10   
324  Melanoma of the Skin             Nevada  0.739669   1.887658e-12   
406              Prostate  Dist. of Columbia  1.322323   8.771763e-12   
143        Uterine Corpus            Alabama  0.763212   3.173587e-13   
293       Lung & Bronchus      West Virginia  1.305448   4.316098e-39   
54         Uterine Cervix         California  1.302660   3.064454e-21   
262       Lung & Bronchus           Kentucky  1.293964   1.896541e-79   
250       Lung & Bronchus           Colorado  0.774074   2.972616e-49   
468       Urinary Bladder              Maine  1.290642   4.037770e-09   
473       Urinary Bladder        Mississippi  0.775782   3.146804e-10   
460       Urinary Bladder             Hawaii  0.779545   2.684345e-05   
459       Urinary Bladder            Georgia  0.782201   5.464066e-27   
70         Uterine Cervix           Michigan  0.783002   3.941284e-06   
146        Uterine Corpus           Arkansas  0.784703   3.206977e-07   
442              Prostate               Utah  1.273881   1.817020e-28   
..                    ...                ...       ...            ...   
479       Urinary Bladder         New Jersey  1.087072   2.649661e-05   
22          Female Breast           Michigan  0.921932   2.981131e-15   
403              Prostate           Colorado  1.083986   7.815007e-08   
251       Lung & Bronchus        Connecticut  0.922914   3.039923e-06   
263       Lung & Bronchus          Louisiana  1.079342   8.648458e-07   
15          Female Breast               Iowa  0.926662   4.107185e-05   
458       Urinary Bladder            Florida  1.075639   4.766980e-08   
444              Prostate           Virginia  1.075229   2.759273e-10   
135        Colon & Rectum              Texas  1.074884   3.996950e-13   
294       Lung & Bronchus          Wisconsin  0.931195   5.087907e-07   
198              Leukemia         California  1.072253   5.084580e-07   
390  Non-Hodgkin Lymphoma              Texas  1.070189   2.126492e-06   
14          Female Breast            Indiana  0.934790   3.566773e-07   
114        Colon & Rectum           Michigan  0.935862   2.344953e-06   
433              Prostate               Ohio  0.935916   9.263466e-12   
147        Uterine Corpus         California  1.065865   4.149715e-06   
277       Lung & Bronchus           New York  0.939370   1.897931e-15   
35          Female Breast               Ohio  0.939577   1.382420e-10   
265       Lung & Bronchus           Maryland  0.941679   2.609148e-05   
42          Female Breast          Tennessee  0.943198   6.072268e-06   
453       Urinary Bladder         California  0.947494   6.047790e-06   
420              Prostate           Michigan  1.055388   3.945941e-08   
266       Lung & Bronchus      Massachusetts  0.948680   4.373935e-05   

Note: the fact that for (Melanoma of the Skin, Texas) we have RR = 0.66 does **NOT mean** that people in Texas are only 66% as likely to develop melanoma compared to other US states. It means that **GIVEN they already have one of the specified types cancer**, it's only 66% as likely to be melanoma (and not one of the other types).